CRM/CiviGeometry/BAO/Geometry.php

```diff
@@ -433,39 +433,147 @@ public static function wkt2kml($wkt) {
+        2 => [$bounds['right_bound'], 'Float'],
+        3 => [$bounds['top_bound'], 'Float'],
+        4 => [$bounds['bottom_bound'], 'Float'],
+        5 => [$geometry_id, 'Integer']
+      ]);
+    } else {
+      CRM_Core_DAO::executeQuery("
+        INSERT INTO $tempTableName (address_id)
+        SELECT ca.id
+        FROM civicrm_address ca
+        WHERE
+              ca.geo_code_2 >= %1
+          AND ca.geo_code_2 <= %2
+          AND ca.geo_code_1 <= %3
+          AND ca.geo_code_1 >= %4
+        ORDER BY ca.id
+      ", [
+        1 => [$bounds['left_bound'], 'Float'],
+        2 => [$bounds['right_bound'], 'Float'],
+        3 => [$bounds['top_bound'], 'Float'],
+        4 => [$bounds['bottom_bound'], 'Float'],
+      ]);
+    }
+
+    $numCandidates = CRM_Core_DAO::singleValueQuery("SELECT COUNT(*) FROM $tempTableName");
+
+    function fetchCandidateBatch($candidateTable, $offset, $batchSize) {
+      $dao = CRM_Core_DAO::executeQuery("
+          SELECT ct.address_id
+          FROM $candidateTable ct
+          ORDER BY address_id
+          LIMIT %1
+          OFFSET %2
+        ", [
+          1 => [$batchSize, 'Integer'],
+          2 => [$offset, 'Integer'],
+        ]);
+
+        $results = $dao->fetchAll();
+
+        return $results;
+    }
+
+    function geometryContainsCandidate($geometryId, $addressId, $geomTableName) {
+      $select = CRM_Utils_SQL_Select::from($geomTableName . ' g, civicrm_address ca')
+        ->select("ST_Contains(g.geometry, ST_GeomFromText(CONCAT('POINT(', ca.geo_code_2, ' ', ca.geo_code_1, ')'), 4326)) AS is_within")
+        ->where("g.id = #geometry_id", ['geometry_id' => $geometryId])
+        ->where("ca.id = #address_id", ['address_id' => $addressId]);
+      $result = CRM_Core_DAO::executeQuery($select->toSQL())->fetchAll();
+
+
+      $result = CRM_Core_DAO::singleValueQuery("
+        SELECT ST_Contains(g.geometry, ST_GeomFromText(CONCAT('POINT(', ca.geo_code_2, ' ', ca.geo_code_1, ')'), 4326))
+        FROM civicrm_address ca, $geomTableName g
+        WHERE g.id = %1 AND ca.id = %2
+      ", [
+        1 => [$geometryId, 'Integer'],
+        2 => [$addressId, 'Integer'],
+      ]);
+
+      return boolval($result);
+    }
+
+    $numBatches = ceil($numCandidates / $params['batch_size']);
+    $offset = 0;
+    $geomTableName = self::getTableName();
+    for ($batchN=1; $batchN <= $numBatches; $batchN++) {
+      $candidates = fetchCandidateBatch($tempTableName, $offset, $params['batch_size']);
+
+      if (count($candidates)) {
+        foreach ($candidates as $candidate) {
+          if (geometryContainsCandidate($geometry_id, $candidate['address_id'], $geomTableName)) {
+            yield ['geometry_id' => $geometry_id, 'address_id' => $candidate['address_id']];
+          }
         }
+
+        $offset += $params['batch_size'];
+      } else {
+        throw new Exception("count <= 0, but haven't finish yet. That doesn't make any sense");
```

Comment:
probably worth while throwing CRM_Core_Execption here instead of the general Exception